### PR TITLE
[IMP] mail,*: config menu for notif settings

### PR DESCRIPTION
**Pull request by zel-odoo**
Add a sub-menu for "Notifications" in the
"Configuration" menu. The "notifications" sub-menu should have the following options to apply to all Discuss channels.
Same options as for notifications on each channel: Mute (+choose time period) / All messages = default / @mentions/ nothing
This setting should apply to all channels that use the default one unless another configuration has been made for a specific one.

split of task-3503901

https://github.com/odoo/enterprise/pull/60535
https://github.com/odoo/upgrade/pull/5909

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/views/discuss/discuss_notification_settings_views.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0"?>
+<odoo>
+    <data>
+        <record id="discuss_notification_settings_action" model="ir.actions.client">
+            <field name="name">Notification Settings</field>
+            <field name="tag">discuss.notification_settings_action</field>
+            <field name="target">new</field>
+            <field name="context">{'dialog_size': 'medium', 'footer': false}</field>
+        </record>
```
Comment (seb-odoo) [inline addons/mail/views/discuss/discuss_notification_settings_views.xml:9]: Confusing split, if the menuitem is in mail (and not discuss folder) this should be too, and so most of the feature too.

The only part that should be in discuss folder should be the channel specific stuff.

File: addons/web/static/src/webclient/actions/action_service.js [hunk 2]
```diff
@@ -923,7 +923,7 @@ export function makeActionManager(env, router = _router) {
             if (size) {
                 actionDialogProps.size = size;
             }
-
+            actionDialogProps.footer = action.context.footer ?? actionDialogProps.footer;
```
Comment (seb-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:926]: What are the fixes in web? Could they be separate PR/fix?

**Review by seb-odoo — COMMENTED** [review]
Review to the end.

This is also very lacking in tests: both python in regards to what we send with web push or what we send on bus, and JS for everything else.

File: addons/mail/static/src/discuss/core/common/discuss_notification_settings.xml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates xml:space="preserve">
+
+    <t t-name="mail.DiscussNotificationSettings">
+        <div class="d-flex flex-column">
+            <div class="d-flex flex-column my-1">
+                <label class="cursor-pointer d-flex align-items-center">
+                    <h5>
+                        <t t-esc="'Mute'"/>
+                    </h5>
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/discuss/core/common/discuss_notification_settings.xml:10]: It should not be a t-esc actually, as it makes the value not translatable.

File: addons/mail/static/src/discuss/core/common/discuss_notification_settings.xml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                    <div class="flex-grow-1"/>
+                    <div class="form-check form-switch">
+                        <input class="form-check-input" type="checkbox" role="switch" t-att-checked="store.settings.mute_until_dt" t-on-change="onChangeDisplayMuteDetails"/>
+                    </div>
+                </label>
+                <div class="d-flex mt-1">
+                    <t t-esc="'Muting a server prevents unread indicators and notifications from appearing.'"/>
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/discuss/core/common/discuss_notification_settings.xml:17]: Also not t-esc

File: addons/mail/static/src/discuss/core/web/messaging_menu_patch.js [hunk 5]
```diff
@@ -16,8 +16,22 @@ patch(MessagingMenu.prototype, {
             this.store.channels.status !== "fetched"
                 ? this.store.initChannelsUnreadCounter
                 : Object.values(this.store.Thread.records).filter(
-                      (thread) => thread.displayToSelf && thread.message_unread_counter > 0
+                      (thread) =>
+                          (thread.displayToSelf &&
+                              !this.store.settings.mute_until_dt &&
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/discuss/core/web/messaging_menu_patch.js:21]: It's not a big deal to check it here too (maybe it's better to always see global check next to thread check!), but then there is an inconsistency with || condition where global mute should also be checked on the case with needaction.

File: addons/mail/static/src/core/common/settings_model.js [hunk 6]
```diff
@@ -38,6 +39,53 @@ export class Settings extends Record {
...
+    /**
+     * General notification settings for channels
+     * @type {'mentions'|'all'|'no_notif'}
+     */
+    channel_notifications = Record.attr("mentions", {
+        compute() {
+            return this.channel_notifications === false ? "mentions" : this.channel_notifications;
+        },
+    });
+    selected_mute_duration = Record.attr(false, {
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/static/src/core/common/settings_model.js:51]: this means which "duration button" is selected so that the chocie is stored locally and when refresh the page it is still there. it is different from the value in the server.

File: addons/mail/static/src/core/web/mail_core_web_service.js [hunk 7]
```diff
@@ -59,6 +59,7 @@ export class MailCoreWeb {
                 if (
                     thread &&
                     message.isNeedaction &&
+                    thread.message_needaction_counter > 0 &&
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/static/src/core/web/mail_core_web_service.js:62]: but functionally it is not working if removed, is it ok with that? i think it is ok for me to remove it and if we have a fix PR for that after freeze it should be ok as well?

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/core/web/mail_core_web_service.js:62]: Yes, remove it and then we can fix, this condition makes no sense I prefer to see the issue.

File: addons/mail/static/src/core/common/settings_model.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/core/common/settings_model.js:51]: I don't understand the point of saving a local value of something that we store on the server.

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/core/common/settings_model.js:51]: It should be there on refresh because the server will give the correct value.

File: addons/mail/static/src/core/common/settings_model.js [hunk 8]
```diff
@@ -38,6 +39,53 @@ export class Settings extends Record {
      * Normalized [0, 1] volume at which the voice activation system must consider the user as "talking".
      */
     voiceActivationThreshold = 0.05;
+    /**
+     * General notification settings for channels
+     * @type {'mentions'|'all'|'no_notif'}
+     */
+    channel_notifications = Record.attr("mentions", {
+        compute() {
+            return this.channel_notifications === false ? "mentions" : this.channel_notifications;
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/static/src/core/common/settings_model.js:48]: it is just more convinient to do so since I checked with @alexkuhn and it is supported in record.

File: addons/mail/static/src/discuss/core/common/notification_settings.xml [hunk 9]
```diff
@@ -7,37 +7,63 @@
...
                         </t>
+                    </Dropdown>
+                </t>
+                <div t-else="" class="d-flex flex-column text-truncate">
+                    <div class="d-flex align-items-center rounded px-3 py-2 m-0">
+                        <span class="fs-6 fw-bold">Mute Channel</span>
+                    </div>
+                    <hr class="solid mx-2 my-0"/>
+                    <t t-foreach="store.settings.MUTES" t-as="item" t-key="item.id">
+                        <button class="o-mail-NotificationSettings-muteDuration btn rounded d-flex align-items-center px-2 py-2 m-0 opacity-75 opacity-100-hover" attrs="{ 'title': item.name }" t-on-click="() => this.setMute(item.value)">
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/static/src/discuss/core/common/notification_settings.xml:40]: it is not coming from the DropdownItem but the Dropdown, so even if it is a DropdownItem it is not having the hovering effects

File: addons/web/static/src/webclient/actions/action_service.js (hunk 2 above)
Comment (zel-odoo) [reply addons/web/static/src/webclient/actions/action_service.js:926]: remove the footer from it. it was not working well previously and paju wants to get rid of the default confirm footer.

File: addons/mail/static/src/discuss/core/common/notification_settings.xml [hunk 10]
```diff
@@ -7,37 +7,63 @@
...
+                            </t>
                         </t>
+                    </Dropdown>
+                </t>
+                <div t-else="" class="d-flex flex-column text-truncate">
+                    <div class="d-flex align-items-center rounded px-3 py-2 m-0">
+                        <span class="fs-6 fw-bold">Mute Channel</span>
+                    </div>
+                    <hr class="solid mx-2 my-0"/>
+                    <t t-foreach="store.settings.MUTES" t-as="item" t-key="item.id">
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/static/src/discuss/core/common/notification_settings.xml:39]: I dont see a lot of repetition, only twice. and each time, the content of the loop is not the same.

[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/static/src/discuss/core/common/notification_settings.xml:39]: actually, there are only two times repetition and each time the content inside is different.

**Comment by zel-odoo** [comment]
Most of the changes were made so that they can be reviewed. Still adding tests and tours

File: addons/test_mail/tests/test_mail_push.py [hunk 11]
```diff
@@ -140,6 +140,13 @@ def test_notify_by_push_channel(self, push_to_end_point):
         push_to_end_point.assert_not_called()
         push_to_end_point.reset_mock()
 
+        self.env["discuss.channel.member"].search([
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_push.py:143]: You dont have default values for mute as far as I know, hence why forcing False ? Those are test users, they have no mute value, do they ?

Comment (zel-odoo) [reply addons/test_mail/tests/test_mail_push.py:143]: the problem is here https://github.com/odoo/odoo/pull/160148/files/f98e0ff5369f7a704e85ebf905abbbfb22b467ad#diff-5f4881b5889644a830bca19abd06b0ee075f57acead3ba2953dd1890ed2f066cR133 is setting the mute so they have a value now.

Comment (zel-odoo) [reply addons/test_mail/tests/test_mail_push.py:143]: L133
```py
self.env['discuss.channel.member'].search([
            ('partner_id', 'in', (self.user_email.partner_id + self.user_inbox.partner_id).ids),
            ('channel_id', 'in', (chat_channel + channel_channel + group_channel).ids),
        ]).write({
            'mute_until_dt': now + timedelta(days=5)
        })
```

File: addons/mail/static/src/core/common/settings_model.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/core/common/settings_model.js:49]: This self-adjusting compute is totally fine here.
I get we could do bad things with self-adjusting computes, but being extremely against them prevents solving issues in the most idiomatic way like here. Who doesn't like simple and reliable code like this one?

File: addons/mail/static/src/discuss/core/common/discuss_notification_settings.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    setup() {
+        this.store = useState(useService("mail.store"));
+    }
+
+    onChangeDisplayMuteDetails() {
+        // set the default mute duration to forever when opens the mute details
+        if (!this.store.settings.mute_until_dt) {
+            this.store.settings.setMuteDuration(
+                this.store.settings.MUTES.find((m) => m.id === "forever").value
+            );
```
Comment (seb-odoo) [inline addons/mail/static/src/discuss/core/common/discuss_notification_settings.js:17]: Ideally "Until I turn it back on" should be selected in the list in this case rather than "select duration".

File: addons/mail/static/src/discuss/core/web/messaging_menu_patch.js [hunk 13]
```diff
@@ -16,8 +16,22 @@ patch(MessagingMenu.prototype, {
             this.store.channels.status !== "fetched"
                 ? this.store.initChannelsUnreadCounter
                 : Object.values(this.store.Thread.records).filter(
-                      (thread) => thread.displayToSelf && thread.message_unread_counter > 0
+                      (thread) =>
+                          thread.displayToSelf &&
+                          !thread.mute_until_dt &&
+                          !this.store.settings.mute_until_dt &&
+                          (thread.message_unread_counter > 0 || thread.message_needaction_counter)
```
Comment (seb-odoo) [inline addons/mail/static/src/discuss/core/web/messaging_menu_patch.js:23]: consistency > 0 or nothing for both

File: addons/mail/models/discuss/discuss_channel.py [hunk 14]
```diff
@@ -549,28 +543,50 @@ def _notify_get_recipients(self, message, msg_vals, **kwargs):
...
+            "active": True,
+            "id": member.partner_id.id,
+            "is_follower": False,
+            "groups": [],
+            "lang": member.partner_id.lang,
+            "notif": "web_push",
+            "share": member.partner_id.partner_share,
+            "type": "customer",
+            "uid": False,
+            "ushare": False,
```
Comment (seb-odoo) [inline addons/mail/models/discuss/discuss_channel.py:588]: miss indent here

File: addons/mail/tests/discuss/test_discuss_channel.py [hunk 15]
```diff
@@ -411,6 +411,121 @@ def test_channel_write_should_send_notification_if_image_128_changed(self):
         ):
             channel.image_128 = base64.b64encode(("<svg/>").encode())
 
+    def test_channel_notification(self):
+        all_test_user = mail_new_test_user(
+            self.env,
+            login='all',
```
Comment (seb-odoo) [inline addons/mail/tests/discuss/test_discuss_channel.py:417]: (minor) lint, prefer double quotes

**Review by seb-odoo — APPROVED** [review]
@robodoo delegate+

**Comment by seb-odoo** [comment]
Also there is an issue on runbot with the new tour apparently.

File: addons/mail/data/ir_cron_data.xml [hunk 16]
```diff
@@ -75,5 +75,14 @@
             <field name="code">model._unmute()</field>
             <field name="state">code</field>
         </record>
+
+        <record id="ir_cron_discuss_users_settings_unmute" model="ir.cron">
```
Comment (KangOl) [inline addons/mail/data/ir_cron_data.xml:79]: Why a new cron? This can be part of the autovacuum cron.

Comment (zel-odoo) [reply addons/mail/data/ir_cron_data.xml:79]: it is not a daily autovacuum cron. it will be called when the user sets a duration for the mute and the cron will be called after the duration.

Comment (KangOl) [reply addons/mail/data/ir_cron_data.xml:79]: Never mind. The cron is triggered as soon as possible explicitly.

**Comment by zel-odoo** [comment]
@robodoo r+

**Comment by zel-odoo** [comment]
@robodoo r+
